### DISCUSSION TITLE
Bluetooth DGT E-board Emulation

### DISCUSSION THREAD

**Discussion by EdNekebno** [EdNekebno/DGTCentaurMods#20] [General]
Some advancement - I have the very very start of getting bluetooth DGT e-board emulation to work:

https://twitter.com/ednekebno/status/1442050920808845313

It's very beginning stages, and of course it takes configuring BT (+rfcomm) etc on the pi - so at some stage I'll look at tidying up my file system on my centaur and making an sd card image.

This works with the Whitepawn app on my tablet, but not yet the DGT Livechess software on my desktop (I started on that but then it very quickly sends an undocumented code). Also whilst in Android the pairing works great, on the PC it's a bit flakey.

The files will get uploaded when I next push. But if you're interested in them sooner and want to drive yourself crazy with bluetooth setup, then let me know and I'll push them quicker.

**Comment by wormstein** [comment]
This is a real breakthrough!! Nice work.
Can you make a branch and push in there what code you have? I'll be happy to help.

-Adrian

**Comment by lucamonfredo** [comment]
That's very promising! Congrats!
I'm in the middle of some reorganisation of the project files in order to tidy up and have some structure.
Hopefully we can then merge everybody's mess (😃) and begin working on a common line.

**Comment by wormstein** [comment]
How did you know how to emulate a DGT e-board? Do you have one?

**Reply by EdNekebno** [reply]
I wish. There's a fair bit of code on github for connecting to DGT e-boards (the bluetooth one just creates a Bluetooth serial port and does the same as the serial one). So there are things like this available: https://github.com/well69/picochess-1/blob/master/test/dgtbrd-ruud.h . Although it doesn't seem to be the latest as, like I say, the Livechess program on windows sends an undocumented message id early on. But in essence it's just a case of writing the server to go with the clients :) The whitepawn app actually only uses very few of the messages. So in theory the process is get that stable and then write the rest in.

**Comment by EdNekebno** [comment]
I'm too unfamiliar with github, so I took the lazy way (I suppose i should work out how to set up a development branch sometime!). I've merged Luca's awesome refactoring and may have hidden pair.py in the config folder and eboard.py in the game folder. Both really really experimental - I just won't add them to the menu until they are a bit more functional.

You need to set up bluetooth and rfcomm - which I forget how I did as I just followed Google/the internet and everything says something different :) Probably easiest to pair directly for now. On my android tablet I've been using WhitePawn and "Chess for Android" to test. You basically start the script, it will insist you position the pieces in the starting position, when it tells you to connect from the other device you cannot the app, and when the centaur LCD reads "Play" you can start moving pieces and experiencing a super buggy experience :)

**Reply by wormstein** [reply]
What I do is this, Still, something is wrong because I cannot see data written to /dev/rfcomm0. I am trying to connect a bt terminal from my phone but faild.

```sudo bluetoothctl discoverable on ```
pair device now

```sudo sdptool browse <btadr>```
Read services on connected device. So communication works. Get channel of Serial Port service

```sudo rfcomm bind 0 <braddr> <channel_no>```
This creates the rfcomm0 device.

LATER EDIT: The steps wor just fine. I managed to communicate between phone and RPi. Use minicom at 115200 8/N/1 and BT Terminal Free app on android.

**Comment by wormstein** [comment]
Do you need to add the board in WhitePawn? How and when you do that?

**Reply by EdNekebno** [reply]
Yep...wait until the centaur LCD says connect remote device or something like that. In WhitePawn click the settings cog and display unknown devices. The board should show on the list and you can select it and then choose the dgt driver.

**Reply by wormstein** [reply]
Got that. Your clip on Twitter was clear enough but rather hard to watch :)))

**Reply by EdNekebno** [reply]
Oh, I always work at an angle like that - stops the blood rushing to my feet! I thought twitter would use the meta data to turn it round...then I couldn't be bothered :)

**Comment by wormstein** [comment]
I wonder how the process of adding the board to whitepwn app works. Have you seen this repo?
https://github.com/wormstein/swpp

I was trying to see what request is passed by the app over serialBT when I try to add the board and select SWPP protocol but I get "Failed to connect" and apparently nothing comes over BT. But I'm sure some request is made... at least according to the protocol documentation some handshake has to happen.

**Reply by EdNekebno** [reply]
I've just tried the SWPP protocol out of interest - but get pretty much the same thing. 

Interesting because it mentions the BLE thing on iphones, so I installed WhitePawn on my iphone rather than my android tablet and it tells me only millenium boards are supported. If they're restricted to BLE on the iphone in general, maybe the developer hasn't given the classic much love. Also a shame - I'd hoped to emulate enough to work from apps on all phones. I just pushed an update to the dtg emulation by the way. i guess we then know the pegasus will be BLE too when it comes.

**Reply by cyberluke** [reply]
I just talk to author of WhitePawn, he might add Pegasus API soon, so then BLE will work on iOS as well.

**Reply by wormstein** [reply]
Not sure if is the same but Ed already broke Pegasus protocol.

https://twitter.com/ednekebno/status/1467110112842158083?t=bO0Hg-M3e_vP2nyvLc8fXg&s=19

**Reply by cyberluke** [reply]
I know, I wrote about author of Whitepawn app - that will be the next step after Ed merge Pegasus feature.

**Comment by wormstein** [comment]
For some reason my board still don't want to connect in WP. Board appears in the list, I select it with DGT Driver but then it simply says Failed to connect and that's it. I suspect there is something I miss on pairing. What's the role of the file in confing?

**Reply by EdNekebno** [reply]
Let me look what my setup is. I'm not sure that's it if you have minicom working, though just in case:

In rc.local I have:

sudo sdptool add SP &

sudo bluetoothctl <<EOF
power on
agent on
discoverable on
pairable on
EOF

I start bluetooth as a service with:

ExecStart=/usr/lib/bluetooth/bluetoothd --compat --noplugin=sap

and afterwards rfcomm with:

ExecStart=/usr/bin/rfcomm watch hci0

The pair.py file in conf runs bt-agent to allow pairing. In /etc/bluetooth/pin.conf I simply have: * 0000
This makes the pin 0000 for everything. For some reason after I've paired this way it remembers it for ever and ever.

If your setup is like mine then there is no /dev/rfcomm0 until something connects. That's important (it's coded in the script as the way it recognises detection).

For debug it is probably best to run eboard from the console. Then in the console it will say: "Setup board"
If all the pieces aren't in the start position it will start to flash lights and also say on the epaper display.
Once all pieces are in the correct place the epaper display will change and the console will read "board is setup"
The epaper then reads "Connect remote device now"

At this point find and connect to the device from the other app (Chess for android is also worth trying as a test, you long hold on the non-board background to bring up a menu).

epaper will refresh the display of the board and read CONNECTED. You can now move pieces.

If you let me know at what point it fails then we can make a guess at what's wrong. 

One thing to keep in mind though is that the version I am working on right now is much more up to date/featured. I just played a game emulating a DGT eboard, through Livechess, on Lichess. (still early days though)

**Reply by wormstein** [reply]
I think this is something we need to take care at setup, isn't?

**Reply by EdNekebno** [reply]
yup...the nightmare bluetooth setup :)

**Comment by OGMalin** [comment]
Just drop in so you know I'm following you.

Here is what the manual say about the 'unknown' command you talk about above:

16 bus command codes from computer to board without data

16.15 DGT_BUS_RPING (0x92)

Purpose:
When multiple boards with the same bus-address are detected, this command can be used to change the bus-address of these boards. When send to a specific address (so not to (0)) all the boards with this address will respond with a message containing a new random address within a random time of 1100ms.

Response:
DGT_MSG_BUS_PING

Note:
The new bus-address will be stored at the board and remains the bus-address until it will be changed again using this command.

**Reply by EdNekebno** [reply]
Ahhhhh....hi, that's really really useful, thanks. So it looks like I should implement more than just a hard coded bus address. Shouldn't be too tough.

It makes no sense in the circumstances that Livechess was sending it to me though. I guess they just use it to tell all boards to pick a new address early on because the chance they would pick the same one is pretty small.

**Reply by OGMalin** [reply]
Looking at some tables I think you can ignore it when using Bluetooth.
Under a table over command that is supported it say 'yes' only on Type-c board (SW V4.x), 'no' on the rest. (Serial, type B, Bluetooth, and Rev-II)
This is true for all commands from 0x91 to 0x97. The last command  Bluetooth support is 0x8d.

**Reply by EdNekebno** [reply]
Good to know. When I ignored it nothing bad happened :) Sounds like you have more information than me. I've been working from https://github.com/well69/picochess-1/blob/master/test/dgtbrd-ruud.h and a bit from the Livechess java source. It doesn't say anything about 0xDF in your information does it? Livechess sends it really early on and it doesn't follow the format of the others. No bus address, no checksum. So not a bus command but a regular one. Again if I ignore it nothing bad happens - so I presumed it's some kind of setup/detection pattern for one of the boards. I haven't yet tracked it down in the Livechess source.

**Reply by OGMalin** [reply]
My version of the manual is from oct. 25, 2019.  Judging from your name of constants it looks like it's from dgtbrd13.h , it was on DGT's web-site earlier, must be about 20 year since I downloaded it (mybe inside a sample program).
I got my copy (95 pages) when asking them from the support page on the web site if there was anything newer, I was strugling with the clock commands. Then the next day it dropped in my mailbox.
I can't find the 0xDF command when take a brief look, it isn't a searchable document. I will come back if I find anything.

**Reply by OGMalin** [reply]
Should have clicked on the link, I see that you use a headerfile from 2009 (edit 2012 or 2014), not the manual itself with explanations.
(It looks like most is covered)

**Comment by wormstein** [comment]
> I start bluetooth as a service with:
>
> ExecStart=/usr/lib/bluetooth/bluetoothd --compat --noplugin=sap
>
> and afterwards rfcomm with:
>
> ExecStart=/usr/bin/rfcomm watch hci0

This part is not very clear. Have you created new services? Put the .service files in etc then so I grab them and install.

**Reply by wormstein** [reply]
Ahhh... I think I get it. You edited `/etc/systemd/system/bluetooth.target.wants/bluetooth.service` and added the rfcomm line as well?

**Reply by EdNekebno** [reply]
Let me check...I can't remember. I did lots of weird stuff :)

**Reply by EdNekebno** [reply]
Okay...there's two services. Both are in /etc/systemd/system.  I put bluetooth.service and rfcomm.service in etc. The bluetooth service was probably created by the apt install and then I modified it.
